org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -149,27 +149,34 @@ public CodeActionHandler(PreferenceManager preferenceManager) {
 	private Optional<Either<Command, CodeAction>> getCodeActionFromProposal(ChangeCorrectionProposal proposal, CodeActionContext context) throws CoreException {
 		String name = proposal.getName();
 
-		Command command;
+		Command command = null;
+		WorkspaceEdit edit = null;
+		boolean useWorkspaceEdit = preferenceManager.getClientPreferences().isWorkspaceApplyEditSupported();
 		if (proposal instanceof CUCorrectionCommandProposal) {
 			CUCorrectionCommandProposal commandProposal = (CUCorrectionCommandProposal) proposal;
 			command = new Command(name, commandProposal.getCommand(), commandProposal.getCommandArguments());
 		} else if (proposal instanceof RefactoringCorrectionCommandProposal) {
 			RefactoringCorrectionCommandProposal commandProposal = (RefactoringCorrectionCommandProposal) proposal;
 			command = new Command(name, commandProposal.getCommand(), commandProposal.getCommandArguments());
 		} else {
-			WorkspaceEdit edit = ChangeUtil.convertToWorkspaceEdit(proposal.getChange());
+			edit = ChangeUtil.convertToWorkspaceEdit(proposal.getChange());
 			if (!ChangeUtil.hasChanges(edit)) {
 				return Optional.empty();
 			}
-			command = new Command(name, COMMAND_ID_APPLY_EDIT, Collections.singletonList(edit));
+			if (!useWorkspaceEdit) {
+				command = new Command(name, COMMAND_ID_APPLY_EDIT, Collections.singletonList(edit));
+			}
 		}
 
 		if (preferenceManager.getClientPreferences().isSupportedCodeActionKind(proposal.getKind())) {
-			// TODO: Should set WorkspaceEdit directly instead of Command
 			CodeAction codeAction = new CodeAction(name);
 			codeAction.setKind(proposal.getKind());
-			codeAction.setCommand(command);
 			codeAction.setDiagnostics(context.getDiagnostics());
+			if (useWorkspaceEdit) {
+				codeAction.setEdit(edit);
```

Comment:
This can easily throw a `NullPointerException`. If the `proposal` is either `CUCorrectionCommandProposal` or `RefactoricCorrectionCommandProposal`, there is *no* `WorkspaceEdit` created.

This is what happened in https://github.com/eclipse/eclipse.jdt.ls/pull/1256#issuecomment-553582214 and was taken care of in https://github.com/eclipse/eclipse.jdt.ls/pull/1278. Good luck.